Mirlo
=====

Overview
--------

Mirlo helps automate the construction
of data sets for multi-gene phylogenetic analyses. It takes whole
proteomes as input, identifies single-copy gene families, aligns the
proteins, and evaluates the 'phylogenetic signal' of each alignment.

What mirlo does:

* Identify single-copy gene families (clusters) in an OrthoFinder report.
* Align each single-copy family using MAFFT
* Construct a phylogenetic tree of each family using PhyML and compute SH-like support values for each branch.
* Compute a 'phylogenetic signal' for each family by computing the mean SH-like values for all branches in each tree. This step is inspired by
Salichos L, and Rokas A: Inferring ancient divergences requires genes
with strong phylogenetic signals. Nature 2013, 497:327â€“331. The authors
show that genes with higher phylogenetic signals have phylogenies that
are more congruent with the species tree. Here, I use SH-like support
values instead of bootstrap values because they are much faster to
compute.

What mirlo does NOT do:

* It does not edit the alignments

* It does not construct a phylogenetic tree from the concatenated
alignment.

Mirlo is a work in progress.

Contact
-------

If you have questions or need help, email me:

Michael Thon mthon@usal.es


Software Requirements
---------------------
- OrthoFinder http://www.stevekellylab.com/software/orthofinder
- ProtTest 3.4 https://code.google.com/p/prottest3/
- PhyML
- mafft
- BioPython


Installation
------------

Uncompress the Mirlo distribution file somewhere on your computer.

Instructions
------------

You should have one file of protein sequences in fasta format for each species that you plan to include in your analysis. Put all of the fasta files into
the one directory. The protein sequence names should be prepended with a short code to identify each species followed by an underscore, for example GLRG_123456. If your sequences don't have this, you can add one on the commend line like this:

`sed 's/>/>GLRG_/' my_proteins.fasta >renamed_proteins.fasta`

just replace GLRG with a unique code for each species.



1. run OrthoFinder.

1. run mirlo.py You can run `python mirlo.py` for more information on the command line parameters. You will need the OrthologousGroups.txt file that was generated by OrthoFinder. Mirlo will output a list of orthologous clusters that have a single gene copy in each species, followed by a value that represents phylogenetic signal. From that list you can decide which orthologous groups you want to include in the concatenated alignment. For example, you may want to use the five clusters with the highest phylogenetic signal.
1. use `cat_alignments.py` to concatenate the selected clusters into one large alignment.
3. The concantenaed alignment can then be used for phylogenetic analysis.



